### PR TITLE
[FLOC-4251] Added instructions on getting started and dependency management.

### PR DESCRIPTION
**Pull request by sarum90**
Fixes https://clusterhq.atlassian.net/browse/FLOC-4251

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+Developing DVOL
+---------------
+
+If you wish to work on DVOL, you'll first need [Go](http://www.golang.org) installed on your machine (version 1.6+ is *required*).
+
+For local dev first make sure Go is properly installed, including setting up a [GOPATH](http://golang.org/doc/code.html#GOPATH). You will also need to add `$GOPATH/bin` to your `$PATH`.
+
+
+Next, using [Git](https://git-scm.com/), clone this repository into `$GOPATH/src/github.com/ClusterHQ/dvol`. All the necessary Golang dependencies are vendored, however you will need to install some Python dependancies and Golang tools -
```
Comment (sarum90) [inline CONTRIBUTING.md]: I'm lazy -- want to add `git clone git@github.com:ClusterHQ/dvol.git $GOPATH/src/github.com/ClusterHQ/dvol` for people like me who just want to copy/paste something?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+```sh
+$ make bootstrap
```
Comment (sarum90) [inline CONTRIBUTING.md]: :question: potentially add a `cd $GOPATH/src/github.com/ClusterHQ/dvol` to the start of this.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+To compile DVOL -
+
+```sh
+$ make build
```
Comment (sarum90) [inline CONTRIBUTING.md]: Hrm...

```
 ♦ /tmp/dvol_test/gopath$ git clone git@github.com:ClusterHQ/dvol.git $GOPATH/src/github.com/ClusterHQ/dvol
Cloning into '/tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol'...
remote: Counting objects: 1518, done.
remote: Compressing objects: 100% (145/145), done.
remote: Total 1518 (delta 77), reused 0 (delta 0), pack-reused 1362
Receiving objects: 100% (1518/1518), 4.97 MiB | 2.45 MiB/s, done.
Resolving deltas: 100% (919/919), done.
Checking connectivity... done.
 ♦ /tmp/dvol_test/gopath$ make bootstrap
make: *** No rule to make target `bootstrap'.  Stop.
 ♦ /tmp/dvol_test/gopath$ ls
bin src
 ♦ /tmp/dvol_test/gopath$ cd src/github.com/ClusterHQ/dvol/
 ♦ /tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol (master)$ ls
Dockerfile         Godeps             LICENSE            Makefile           README.md          TODO.md            demos              docker-compose.yml dvol.go            dvol_python        install.sh         pkg                rebuild.sh         scripts            setup.py           vendor
 ♦ /tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol (master)$ make bootstrap
go get github.com/tools/godep
go get golang.org/x/tools/cmd/cover
go get golang.org/x/tools/cmd/vet
go get golang.org/x/tools/cmd/goimports
go get github.com/pierrre/gotestcover
test -d venv || virtualenv venv
New python executable in venv/bin/python2.7
Also creating executable in venv/bin/python
Installing setuptools, pip, wheel...done.
. venv/bin/activate \
	&& pip install .
You are using pip version 7.1.0, however version 8.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Processing /private/tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol
Collecting PyYAML>=3 (from dvol==0.1)
Collecting Twisted>=14.0.2 (from dvol==0.1)
Collecting treq>=14 (from dvol==0.1)
  Using cached treq-15.1.0-py2.py3-none-any.whl
Collecting pyasn1>=0.1 (from dvol==0.1)
  Using cached pyasn1-0.1.9-py2.py3-none-any.whl
Collecting hypothesis>=2.0 (from dvol==0.1)
  Downloading hypothesis-3.0.5.tar.gz (59kB)
    100% |████████████████████████████████| 61kB 697kB/s 
Collecting docker-py>=1.5.0 (from dvol==0.1)
  Downloading docker-py-1.7.2.tar.gz (68kB)
    100% |████████████████████████████████| 69kB 748kB/s 
Collecting zope.interface>=3.6.0 (from Twisted>=14.0.2->dvol==0.1)
Collecting service-identity>=14.0.0 (from treq>=14->dvol==0.1)
  Downloading service_identity-16.0.0-py2.py3-none-any.whl
Collecting pyOpenSSL>=0.13 (from treq>=14->dvol==0.1)
  Using cached pyOpenSSL-0.15.1-py2.py3-none-any.whl
Collecting requests>=2.1.0 (from treq>=14->dvol==0.1)
  Using cached requests-2.9.1-py2.py3-none-any.whl
Collecting enum34 (from hypothesis>=2.0->dvol==0.1)
Collecting six>=1.4.0 (from docker-py>=1.5.0->dvol==0.1)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting websocket-client>=0.32.0 (from docker-py>=1.5.0->dvol==0.1)
Requirement already satisfied (use --upgrade to upgrade): setuptools in ./venv/lib/python2.7/site-packages (from zope.interface>=3.6.0->Twisted>=14.0.2->dvol==0.1)
Collecting pyasn1-modules (from service-identity>=14.0.0->treq>=14->dvol==0.1)
  Using cached pyasn1_modules-0.0.8-py2.py3-none-any.whl
Collecting attrs (from service-identity>=14.0.0->treq>=14->dvol==0.1)
  Using cached attrs-15.2.0-py2.py3-none-any.whl
Collecting cryptography>=0.7 (from pyOpenSSL>=0.13->treq>=14->dvol==0.1)
  Downloading cryptography-1.2.2-cp27-none-macosx_10_6_intel.whl (1.4MB)
    100% |████████████████████████████████| 1.4MB 268kB/s 
Collecting ipaddress (from cryptography>=0.7->pyOpenSSL>=0.13->treq>=14->dvol==0.1)
  Using cached ipaddress-1.0.16-py27-none-any.whl
Collecting idna>=2.0 (from cryptography>=0.7->pyOpenSSL>=0.13->treq>=14->dvol==0.1)
  Using cached idna-2.0-py2.py3-none-any.whl
Collecting cffi>=1.4.1 (from cryptography>=0.7->pyOpenSSL>=0.13->treq>=14->dvol==0.1)
  Downloading cffi-1.5.2.tar.gz (388kB)
    100% |████████████████████████████████| 389kB 897kB/s 
Collecting pycparser (from cffi>=1.4.1->cryptography>=0.7->pyOpenSSL>=0.13->treq>=14->dvol==0.1)
Building wheels for collected packages: hypothesis, docker-py, cffi
  Running setup.py bdist_wheel for hypothesis
  Stored in directory: /Users/mewert/Library/Caches/pip/wheels/04/cf/56/72edbd84c21f5a284d5a657ed8fea4f1091d50e6d6e1f38027
  Running setup.py bdist_wheel for docker-py
  Stored in directory: /Users/mewert/Library/Caches/pip/wheels/0a/26/d0/6acfa391aeb1824475d9125c88d075ae9b66e16107961f6678
  Running setup.py bdist_wheel for cffi
  Stored in directory: /Users/mewert/Library/Caches/pip/wheels/27/94/68/1a8c5c1e8492c8a853e5c96e29fef5a62d00230cefa1389eb5
Successfully built hypothesis docker-py cffi
Installing collected packages: PyYAML, zope.interface, Twisted, six, enum34, ipaddress, pyasn1, idna, pycparser, cffi, cryptography, pyOpenSSL, pyasn1-modules, attrs, service-identity, requests, treq, hypothesis, websocket-client, docker-py, dvol
  Running setup.py install for dvol
Successfully installed PyYAML-3.11 Twisted-15.5.0 attrs-15.2.0 cffi-1.5.2 cryptography-1.2.2 docker-py-1.7.2 dvol-0.1 enum34-1.1.2 hypothesis-3.0.5 idna-2.0 ipaddress-1.0.16 pyOpenSSL-0.15.1 pyasn1-0.1.9 pyasn1-modules-0.0.8 pycparser-2.14 requests-2.9.1 service-identity-16.0.0 six-1.10.0 treq-15.1.0 websocket-client-0.35.0 zope.interface-4.1.3
 ♦ /tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol (master)$ make build
godep go build .
pkg/cmd/root.go:8:2: cannot find package "github.com/spf13/cobra" in any of:
	/usr/local/Cellar/go/1.5.1/libexec/src/github.com/spf13/cobra (from $GOROOT)
	/tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol/Godeps/_workspace/src/github.com/spf13/cobra (from $GOPATH)
	/tmp/dvol_test/gopath/src/github.com/spf13/cobra
godep: go exit status 1
make: *** [build] Error 1
 ♦ /tmp/dvol_test/gopath/src/github.com/ClusterHQ/dvol (master)$
```

Comment (sarum90) [inline CONTRIBUTING.md]: maybe it's a problem with homebrew go?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+```
+
+To run the tests -
+
+```sh
+$ make test
```
Comment (sarum90) [inline CONTRIBUTING.md]: :question: so much text! Do we have plans to reroute that somewhere else? or is it helpful for debugging?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
+Developing DVOL
+---------------
+
+If you wish to work on DVOL, you'll first need [Go](http://www.golang.org) installed on your machine (version 1.6+ is *required*).
+
+For local dev first make sure Go is properly installed, including setting up a [GOPATH](http://golang.org/doc/code.html#GOPATH). You will also need to add `$GOPATH/bin` to your `$PATH`.
```
Comment (sarum90) [inline CONTRIBUTING.md]: :question: If you want to enable my laziness: `export PATH=$PATH:$GOPATH/bin`

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+If you're developing DVOL, there are a few tasks you might need to perform.
+
+#### Adding a dependency
+
+If you're adding a dependency. You'll need to vendor it in the same Pull Request as the code that depends on it. You should do this in a separate commit from your code, as makes PR review easier and Git history simpler to read in the future.
+
+Because godep captures new dependencies from the local `$GOPATH`, you first need to `godep restore` from the master branch to ensure that the only diff is your new dependency.
+
+Assuming your work is on a branch called `my-feature-branch`, the steps look like this:
```
Comment (sarum90) [inline CONTRIBUTING.md]: Here is "your work":

- a pending branch that adds code that includes a dependency you `go get`'d?
- just a pending branch modified by a `go get`?
- a pending branch in which you manually downloaded something into vendor/?
- a pending branch that adds code that includes a dependency of something you added to vendor/?
- something else?

Trying to figure out in order to verify instructions.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+git checkout master
+git pull
+godep restore -v # flag is optional, enables verbose output
+
+# Capture the new dependency referenced from my-feature-branch
+git checkout my-feature-branch
+git rebase master
+godep save ./...
+
+# There should now be changes in `vendor/` with added files for your dependency,
```
Comment (sarum90) [inline CONTRIBUTING.md]: :question: I tried:

-Making a branch that included a new github dependency.
-Running the code failed so I ran `go get dependency`
-Running the code worked, so then I committed the state of the world and started these directions.

At this point I only had 1 change (other than importing the dependency and some .gitignore slopiness (venv folder and cover.out).
```
--- a/Godeps/Godeps.json
+++ b/Godeps/Godeps.json
@@ -1,12 +1,20 @@
 {
-       "ImportPath": "github.com/clusterhq/dvol",
+       "ImportPath": "github.com/ClusterHQ/dvol",
        "GoVersion": "go1.6",
+       "Packages": [
+               "./..."
+       ],
        "Deps": [
                {
                        "ImportPath": "github.com/inconshreveable/mousetrap",
                        "Rev": "76626ae9c91c4f2a10f34cad8ce83ea42c93bb75"
                },
                {
+                       "ImportPath": "github.com/jordan-wright/email",
+                       "Comment": "v2.1-10-ga62870b",
+                       "Rev": "a62870b0c368c3c29de27027be67d855e308cc37"
+               },
+               {
                        "ImportPath": "github.com/spf13/cobra",
                        "Rev": "65a708cee0a4424f4e353d031ce440643e312f92"
                },
```

I'm not sure what I should have done to get the vendor/ change.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+# and changes in Godeps/Godeps.json with metadata for your dependency.
+
+# Make a commit with your new dependencies added
+git add -A
```
Comment (sarum90) [inline CONTRIBUTING.md]: I think we might want to add more to the local .gitignore before encouraging people to do git add -A

specifically:
```
/cover.out
/dvol
/venv/
```

**Comment by sarum90** [comment]
Thanks!

It's nice to have this procedure written down. So I'm not super familiar with `godep` so I had some troubles figuring out precisely how to add a new dependency and have it show up in `vendor/`

Otherwise I've got a few nits throughout. Take them or not at your own discretion, but this looks fine to merge.

**Comment by davidoliverSP2** [comment]
Have pushed a few updates to the docs. If it goes through any further revisions that you'd like me to review, please ping. Otherwise, looks good - thanks @mdevilliers & @sarum90

**Pull request merged**